File: src/Matriz/CGNE.java
package Matriz;
import org.jblas.DoubleMatrix;

public class CGNE {
    public int iteractions = 0;

    public DoubleMatrix getImage(DoubleMatrix H, DoubleMatrix g, Integer selectedModelFlag){
        System.out.println("H rows: " + H.rows + " x " + H.columns + " and g rows: " + g.columns);

        DoubleMatrix f = DoubleMatrix.zeros(H.columns);
        
        DoubleMatrix r = g.sub(H.mmul(f));
        
        DoubleMatrix p = H.transpose().mmul(r);

        int i = 0;

        do{
            DoubleMatrix alpha = r.transpose().mmul(r).div(
                p.transpose().mmul(p)
            );

            DoubleMatrix f1 = f.add(alpha.mmul(p));
            
            DoubleMatrix r1 = r.sub(alpha.mmul(H.mmul(p)));
            
            DoubleMatrix beta = r1.transpose().mmul(r1).div(
                r.transpose().mmul(r)
            );

            DoubleMatrix p1 = H.transpose().mmul(r1).add(beta.mmul(p));

            f = f1;
            r = r1;
            p = p1;
            i++;

        }while(i < 30);
        
        if (selectedModelFlag == 1) {
            f.reshape(60, 60);
        }
        else {
            f.reshape(30, 30);
        }

        return f;
    }
}

File: CGNE.py
import org.jblas.DoubleMatrix

class CGNE:
    def __init__(self):
        self.iterations = 0

    def get_image(self, h, g, selected_model_flag):
        f = DoubleMatrix.zeros(h.columns)
        r = g - h.mmul(f)
        z = h.transpose().mmul(r)

        p = z
        error = float('inf')
        i = 0
        while error > 1e-4 and i < 30:
            alpha = r.transpose().mmul(r) / p.transpose().mmul(p)
            new_f = f + alpha.mmul(p)
            new_r = r - alpha.mmul(h.mmul(p))
            beta = new_r.transpose().mmul(new_r) / r.transpose().mmul(r)
            new_p = h.transpose().mmul(new_r) + beta.mmul(p)

            error = self.error(new_r, r)

            f = new_f
            r = new_r
            p = new_p
            i += 1

        self.iterations = i

        if selected_model_flag == 1:
            f.reshape(60, 60)
        else:
            f.reshape(30, 30)

        return f

    def get_iterations(self):
        return self.iterations

    def error(self, new_r, r):
        return (new_r - r).norm2()